core/src/cluster_slots_service/cluster_slots.rs

```diff
@@ -221,142 +337,111 @@ impl ClusterSlots {
                 debug_assert_eq!(*s, slot, "Fetched slot does not match expected value!");
-                let slot_weight_f64 =
-                    total_support.load(std::sync::atomic::Ordering::Relaxed) as f64;
-                // once slot is confirmed beyond `FREEZE_THRESHOLD` we should not need
-                // to update the stakes for it anymore
-                if slot_weight_f64 / total_stake as f64 > FREEZE_THRESHOLD {
+                if map.is_frozen() {
                     continue;
                 }
-                let rg = map.read().unwrap();
-                let committed_stake = rg.get(&epoch_slots.from);
-                // if there is already a record, we can atomic swap a correct value there
-                if let Some(committed_stake) = committed_stake {
-                    let old_stake = committed_stake.swap(sender_stake, Ordering::Relaxed);
-                    if old_stake == 0 {
-                        total_support.fetch_add(sender_stake, Ordering::Relaxed);
-                    }
-                } else {
-                    drop(rg);
-                    {
-                        let mut wg = map.write().unwrap();
-                        wg.insert(epoch_slots.from, AtomicU64::new(sender_stake));
-                    }
-                    self.metric_write_locks.fetch_add(1, Ordering::Relaxed);
+                if map
+                    .set_support_by_pubkey(&epoch_slots.from, sender_stake)
+                    .is_err()
+                {
+                    error!("Unexpected pubkey {} for slot {}!", &epoch_slots.from, slot);
+                    break;
                 }
             }
         }
     }
 
-    /// Returns number of stored entries
-    fn datastructure_size(&self) -> usize {
-        let cluster_slots = self.cluster_slots.read().unwrap();
-        cluster_slots
-            .iter()
-            .map(|RowContent { supporters, .. }| supporters.read().unwrap().len())
-            .sum::<usize>()
-    }
-
-    fn report_cluster_slots_perf_stats(&self) {
+    /// Upload stats into metrics database
+    fn maybe_report_cluster_slots_perf_stats(&self) {
         if self.metrics_last_report.should_update(10_000) {
             let write_locks = self.metric_write_locks.swap(0, Ordering::Relaxed);
             let allocations = self.metric_allocations.swap(0, Ordering::Relaxed);
+            let cluster_slots = self.cluster_slots.read().unwrap();
+            let (size, frozen, blank) =
+                cluster_slots
+                    .iter()
+                    .fold((0, 0, 0), |(s, f, b), RowContent { supporters, .. }| {
+                        (
+                            s + supporters.memory_usage(),
+                            f + supporters.is_frozen() as usize,
+                            b + supporters.is_blank() as usize,
+                        )
+                    });
             datapoint_info!(
                 "cluster-slots-size",
```

Comment:
Report blanks and frozen entries separately to make sure we can keep track of things inside this datastructure